xmsannotator/R/compute_scores.R

```diff
@@ -1,35 +1,157 @@
+                            peaks,
+                            abundance_ratio,
+                            mass_defect_tolerance,
+                            rt_tolerance) {
   isotopes <- mutate(
     peaks,
+    mass_number_difference = round(mz - query$expected_mass),
     compound = query$compound,
-    isotope_number = round(mz - query$monoisotopic_mass),
-    adduct = sprintf("%s_[%+d]", query$adduct, isotope_number),
-    formula = sprintf("%s_[%+d]", query$formula, isotope_number),
+    adduct = query$adduct,
+    molecular_formula = query$molecular_formula
   )
   isotopes <- filter(
     isotopes,
-    cluster == query$cluster,
-    avg_intensity / query$avg_intensity <= abudance_ratio,
+    rt_cluster == query$rt_cluster,
+    mean_intensity / query$mean_intensity <= abundance_ratio + abundance_ratio * intensity_deviation_tolerance,
     near(rt, query$rt, rt_tolerance),
-    near(mass_defect, query$mass_defect, mass_defect_tolerance),
-    between(abs(isotope_number), 1, max_isp)
+    near(mass_defect, query$mass_defect, mass_defect_tolerance)
   )
-  isotopes <- select(isotopes, -isotope_number)
 }
 
+#' Match prefiltered isotopes by comparing their relative intensity to their relative natural abundance.
+#' The relative values are computed as a ratio of an isotopic peak intensity to the intensity of its monoisotopic peak,
+#' respectively as a ratio of a natural abundance of the isotope to the abundance of the most abundant isotope.
+#'
+#' @param query A single annotated peak.
+#' @param isotopes Isotope candidates for the annotated molecule. The candidates are identified in [compute_isotopes].
+#' @param pattern Isotopic pattern of a given molecule as produced by [compute_isotopic_pattern].
+#' @param intensity_deviation_tolerance A numeric threshold by which an intensity ratio of two isotopic peaks may differ
+#'  from their actual abundance ratio.
+#'
+#' @return A table of matched isotopes.
+#'
+#' @import dplyr
+match_isotopes_by_intensity <- function(query,
+                                        isotopes,
+                                        pattern,
+                                        intensity_deviation_tolerance) {
+  isotopes <- mutate(isotopes,
+                     relative_intensity = mean_intensity / query$mean_intensity)
+  isotopes <- left_join(isotopes,
+                        select(pattern, mass_number_difference, abund),
+                        by = "mass_number_difference")
+  isotopes <- filter(isotopes,
+                     near(relative_intensity, abund, relative_intensity * intensity_deviation_tolerance))
+  isotopes <- select(isotopes, -c(abund, relative_intensity))
+}
+
+#' Find all possible isotopes of a given annotated peak by filtering peaks based on several criteria
+#' (see [filter_isotopes]) and matching the intensities with natural abundances of the isotopes.
+#'
+#' @param ... A single isotope passed as a list of columns. See [purrr::map()] for more details.

```

Comment:
'a single isotope passed as a list of columns' - isn't it more a single annotated feature (row from annotation table) for which we want to detect the isotopic peaks in the peak table?

Reply:
Resolved in c3bc230e362406a8856b79f522e986ef2b502d7c